changelog.py

```diff
@@ -0,0 +1,168 @@
+import gitlab
+import os.path
+import re
+import semver
+import yaml
+
+from dateutil import parser
+from datetime import datetime, timedelta, timezone
+from jinja2 import Template
+from typing import Dict
+from urllib.parse import unquote
+
+
+path = os.path.join(os.path.dirname(__file__), "template.yml")
+with open(path) as f:
+    action = yaml.safe_load(f)
+DEFAULT_TEMPLATE = action["changelog"]["default"]
+
+DEFAULT_IGNORE = [
+    "changelog skip",
+    "duplicate",
+    "exclude from changelog",
+    "invalid",
+    "no changelog",
+    "question",
+    "wont fix",
+]
+
+
+class Changelog:
+    """A Changelog produces release notes for a single release."""
+
+    def __init__(self, repo):
+        self._repo = repo
+        self._template = Template(DEFAULT_TEMPLATE, trim_blocks=True)
+        self._ignore = set(self._slug(s) for s in DEFAULT_IGNORE)
+
+    def _slug(self, s: str) -> str:
+        """Return a version of the string that's easy to compare."""
+        return re.sub(r"[\s_-]", "", s.casefold())
```

Comment:
Could potentially pre-compile the regex in the init if it's going to be used multiple times.
Put this in init: `self._slug_re = re.compile(r"[\s_-]")`
and then here you would do the following in this function: `return self._slug_re.sub("", s.casefold())`

Not really sure how much processing this saves, but if a regex is going to happen multiple times I like to compile it.